### PR TITLE
[5.1] Fix bug with expectsEvents() when events are fired directly on dispatcher.

### PR DESCRIPTION
**Pull request by reinink**
Okay, so this isn't a simple bug and requires a little explanation. When writing tests the `expectsEvents()` method allows you to verify that certain events are called. This automatically mocks the event dispatcher, and verifies that each event is called at least once. However, there is a slight problem with how Mockery verifies this, where the optional `fire()` method parameters must actually be present, otherwise the assertion will fail.

The result is that unless an event is fired with the `event()` helper function, these tests will always fail. This is because the `event()` helper function explicitly sets `$payload` and `$halt` optional parameters (as seen [here](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Foundation/helpers.php#L658)), but they are (naturally) optional when working directly with the dispatcher object.

Some examples:

``` php
// Passes:
event(new MyEvent($model));
Event::fire(new MyEvent($model), [], false);
$dispatcher->fire(new MyEvent($model), [], false);

// Fails:
Event::fire(new MyEvent($model));
$dispatcher->fire(new MyEvent($model));
```

I've [spoken](https://github.com/padraic/mockery/issues/476) with the folks over at Mockery, and there currently isn't a way for Mockery to specify these as "optional, optional parameters". What you see in this PR is a workaround that fixes this problem.

Basically what's happening is I'm manually tracking each event that's fired using a `$mock->shouldReceive()` callback. Then after the test completes, I analyze what events were fired. If an expected event doesn't appear in the `$calledEvents` array, an exception is thrown.

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/ApplicationTrait.php [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 namespace Illuminate\Foundation\Testing;
 
+use Exception;
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Please sort imports by length.

File: src/Illuminate/Foundation/Testing/ApplicationTrait.php [hunk 2]
```diff
@@ -55,18 +56,24 @@ protected function instance($abstract, $instance)
      * @param  array|dynamic  $events
      * @return $this
      */
-    public function expectsEvents($events)
+    public function expectsEvents($expectedEvents)
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Please update the phpdoc.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Also, "$events" was fine to begin with.

File: src/Illuminate/Foundation/Testing/ApplicationTrait.php (hunk 1 above)
Comment (reinink) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Will do. (I have PHP fixer setup to sort by name automatically.)

**Comment by reinink** [comment]
Okay, I've made the requested changes. Note, I've actually improved this bug fix, as it now also accounts for events fired as strings, interfaces and subclasses. Further, if multiple expected events are not fired, it will show them all in the exception message, not just the first one in the list.

File: src/Illuminate/Foundation/Testing/ApplicationTrait.php [hunk 3]
```diff
@@ -59,14 +60,25 @@ public function expectsEvents($events)
...
-
-        foreach ($events as $event) {
-            $mock->shouldReceive('fire')->atLeast()->once()
-                ->with(Mockery::type($event), [], false);
-        }
+        $mock = Mockery::spy('Illuminate\Contracts\Events\Dispatcher');
+
+        $mock->shouldReceive('fire')->andReturnUsing(function ($called) use (&$events) {
+            foreach ($events as $key => $event) {
+                if ((is_string($called) and $called === $event) or
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Please use `&&` and `||`.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: In all places.

Comment (reinink) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Really? Who even uses those anymore? :-)

Comment (reinink) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: I know, I know, consistency. Just bugging.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: We actually had them changed from "and" to "&&" everywhere. :)

Comment (reinink) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Out of genuine curiosity, why?

**Comment by reinink** [comment]
Updated `or` to `||` and `and` to `&&`.

File: src/Illuminate/Foundation/Testing/ApplicationTrait.php (hunk 3 above)
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: Because `&&` normally suits us more than `and`. They're not quite the same. ;)

**Comment by GrahamCampbell** [comment]
This looks good to me. :+1:

**Comment by tomturton** [comment]
Has this fix been released yet? I'm still getting this bug with 5.1.20/

**Comment by GrahamCampbell** [comment]
This commit is included in all releases from v5.1.2 onwards.

**Comment by raftalks** [comment]
I am experiencing the same issue on the dev branch for laravel 5.5 as well, any plans to fix this?
